ssi-ldp/src/suites/w3c.rs

```diff
@@ -0,0 +1,593 @@
+        let key = resolve_key(verification_method, resolver).await?;
+        self.validate_key_and_algorithm(&key, header.algorithm)?;
+        Ok(ssi_jws::verify_bytes_warnable(
+            header.algorithm,
+            &signing_input,
+            &key,
+            &signature,
+        )?)
+    }
+    async fn complete(
+        &self,
+        preparation: ProofPreparation,
+        signature: &str,
+    ) -> Result<Proof, Error> {
+        complete(preparation, signature).await
+    }
+}
+
+impl JsonWebSignature2020 {
+    fn validate_algorithm(&self, algorithm: Algorithm) -> Result<(), Error> {
+        match algorithm {
+            Algorithm::EdDSA => (),
+            Algorithm::ES256K => (),
+            Algorithm::ES256 => (),
+            Algorithm::ES384 => (),
+            Algorithm::PS256 => (),
+            _ => return Err(Error::JWS(ssi_jws::Error::UnsupportedAlgorithm)),
+        }
+        Ok(())
+    }
+    // https://w3c-ccg.github.io/lds-jws2020/#jose-conformance
+    fn validate_key_and_algorithm(&self, key: &JWK, algorithm: Algorithm) -> Result<(), Error> {
+        if let Some(key_algorithm) = key.algorithm {
+            if key_algorithm != algorithm {
+                return Err(Error::JWS(ssi_jws::Error::AlgorithmMismatch));
+            }
+        }
+        match &key.params {
+            JWKParams::RSA(_) => match algorithm {
+                Algorithm::PS256 => (),
+                _ => return Err(Error::JWS(ssi_jws::Error::UnsupportedAlgorithm)),
+            },
+            JWKParams::EC(ec_params) => {
+                match &ec_params
+                    .curve
+                    .as_ref()
+                    .ok_or(ssi_jwk::Error::MissingCurve)?[..]
+                {
+                    "secp256k1" => match algorithm {
+                        Algorithm::ES256K => (),
+                        _ => return Err(Error::JWS(ssi_jws::Error::UnsupportedAlgorithm)),
+                    },
+                    "P-256" => match algorithm {
+                        Algorithm::ES256 => (),
+                        _ => return Err(Error::JWS(ssi_jws::Error::UnsupportedAlgorithm)),
+                    },
+                    "P-384" => match algorithm {
+                        Algorithm::ES384 => (),
+                        _ => return Err(Error::JWS(ssi_jws::Error::UnsupportedAlgorithm)),
+                    },
+                    _ => {
+                        return Err(Error::UnsupportedCurve);
+                    }
+                }
+            }
+            JWKParams::OKP(okp_params) => match &okp_params.curve[..] {
+                "Ed25519" => match algorithm {
+                    Algorithm::EdDSA => (),
+                    _ => return Err(Error::JWS(ssi_jws::Error::UnsupportedAlgorithm)),
+                },
+                _ => {
+                    return Err(Error::UnsupportedCurve);
+                }
+            },
+            _ => {}
```

Comment:
This empty case is not an error? I'm assuming that if it's `Symmetric` there is nothing to check?

Reply:
good point, we don't have any corresponding `Algorithm` variants but I suppose that means we should return `UnsupportedAlgorithm` or something